File: problems/322_coin-change/main.cpp
#include <iostream>
#include <vector>

class Solution
{
public:
    int coinChange(std::vector<int> &coins, int amount)
    {
        int INF = amount + 1;
        std::vector<int> dp(amount + 1, INF);
        dp[0] = 0;
        for (auto c : coins)
        {
            for (int i = c; i <= amount; i++)
                dp[i] = std::min(dp[i], dp[i - c] + 1);
        }
        return dp[amount] == INF ? -1 : dp[amount];
    }

    int rec(std::vector<int> &coins, int amt, std::vector<int> &dp)
    {
        if (amt == 0)
            return 0;
        if (amt < 0)
            return INT_MAX / 2; // safe INF

        if (dp[amt] != -1)
            return dp[amt]; // memo hit

        int count = INT_MAX / 2;
        for (int c : coins)
        {
            int res = rec(coins, amt - c, dp);
            count = std::min(count, res + 1);
        }

        dp[amt] = count; // memo store
        return count;
    }

    int coinChange_2(std::vector<int> &coins, int amount)
    {
        std::vector<int> dp(amount + 1, -1); // -1 = uncomputed
        int res = rec(coins, amount, dp);
        return (res >= INT_MAX / 2) ? -1 : res;
    }
};

int main()
{
    Solution solution;
    std::vector<std::pair<std::pair<std::vector<int>, int>, int>> samples = {
        {{{1, 2, 5}, 11}, 3},
        {{{2}, 3}, -1},
        {{{1}, 0}, 0},
    };

    for (const auto &sample : samples)
    {
        std::vector<int> coins = sample.first.first;
        int amount = sample.first.second;
        int expected = sample.second;
        int result = solution.coinChange(coins, amount);

        std::cout << "Input: coins = [";
        for (size_t i = 0; i < coins.size(); ++i)
        {
            std::cout << coins[i];
            if (i + 1 < coins.size())
            {
                std::cout << ",";
            }
        }
        std::cout << "], amount = " << amount << '\n';
        std::cout << "Placeholder result: " << result << '\n';
        std::cout << "Expected output: " << expected << "\n\n";
    }

    return 0;
}

File: problems/322_coin-change/main.rs
struct Solution;

impl Solution {
    pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
        if amount == 0 {
            return 0;
        }
        let inf = amount + 1;
        let mut dp = vec![inf; (amount + 1) as usize];
        dp[0] = 0;
        for &c in &coins {
            for i in c..=amount {
                dp[i as usize] = dp[i as usize].min(dp[(i - c) as usize] + 1);
            }
        }
        if dp[amount as usize] == inf {
            -1
        } else {
            dp[amount as usize]
        }
    }

    pub fn coin_change_2(coins: Vec<i32>, amount: i32) -> i32 {
        let mut dp = vec![-1; (amount + 1) as usize];
        let res = Self::rec(&coins, amount, &mut dp);
        if res >= i32::MAX / 2 {
            -1
        } else {
            res
        }
    }

    fn rec(coins: &Vec<i32>, amt: i32, dp: &mut Vec<i32>) -> i32 {
        if amt == 0 {
            return 0;
        }
        if amt < 0 {
            return i32::MAX / 2;
        }
        if dp[amt as usize] != -1 {
            return dp[amt as usize];
        }
        let mut count = i32::MAX / 2;
        for &c in coins {
            let res = Self::rec(coins, amt - c, dp);
            count = count.min(res + 1);
        }
        dp[amt as usize] = count;
        count
    }
}

fn main() {
    let samples = vec![(vec![1, 2, 5], 11, 3), (vec![2], 3, -1), (vec![1], 0, 0)];

    for (coins, amount, expected) in samples {
        let result = Solution::coin_change(coins.clone(), amount);
        println!("Input: coins = {:?}, amount = {}", coins, amount);
        println!("Placeholder result: {result}");
        println!("Expected output: {expected}\n");
    }
}
